e2e/workload_id/nodemeta_test.go

```diff
@@ -0,0 +1,139 @@
+func testDynamicNodeMetadata(t *testing.T) {
+	nomad := e2eutil.NomadClient(t)
+
+	nodes, _, err := nomad.Nodes().List(nil)
+	must.NoError(t, err)
+	if len(nodes) == 0 {
+		t.Skip("requires at least 1 linux node")
+	}
+
+	node := nodes[0]
+
+	keyFoo := "foo-" + uuid.Short()
+	keyEmpty := "empty-" + uuid.Short()
+	keyUnset := "unset-" + uuid.Short()
+
+	// Go ahead and submit job so it is scheduled as soon as the node metadata is
+	// applied
+	jobID := "node-meta-" + uuid.Short()
+	jobIDs := []string{jobID}
+	t.Cleanup(e2eutil.CleanupJobsAndGC(t, &jobIDs))
+
+	path := "./input/node-meta.nomad.hcl"
+	jobBytes, err := os.ReadFile(path)
+	must.NoError(t, err)
+	job, err := jobspec2.ParseWithConfig(&jobspec2.ParseConfig{
+		Path: path,
+		Body: jobBytes,
+		ArgVars: []string{
+			"foo_constraint=${meta." + keyFoo + "}",
+			"empty_constraint=${meta." + keyEmpty + "}",
+			"unset_constraint=${meta." + keyUnset + "}",
+			"foo_key=" + keyFoo,
+			"empty_key=" + keyEmpty,
+			"unset_key=" + keyUnset,
+		},
+		Strict: true,
+	})
+	must.NoError(t, err)
+	job.ID = pointer.Of(jobID)
+	_, _, err = nomad.Jobs().Register(job, nil)
+	must.NoError(t, err)
+
+	// Update the node meta to allow the job to be placed
+	req := &api.NodeMetaApplyRequest{
+		NodeID: node.ID,
+		Meta: map[string]*string{
+			keyFoo:   pointer.Of("bar"),
+			keyEmpty: pointer.Of(""),
+			keyUnset: nil,
+		},
+	}
+	resp, err := nomad.Nodes().Meta().Apply(req, nil)
+	must.NoError(t, err)
+	must.Eq(t, "bar", resp.Meta[keyFoo])
+	must.MapContainsKey(t, resp.Meta, keyEmpty)
+	must.Eq(t, "", resp.Meta[keyEmpty])
+	must.MapNotContainsKey(t, resp.Meta, keyUnset)
+
+	// Wait up to 10 seconds (with 1s buffer) for updates to be visible to the
+	// scheduler.
+	qo := &api.QueryOptions{
+		AllowStale: false,
+		WaitIndex:  node.ModifyIndex,
+		WaitTime:   2 * time.Second,
+	}
+	deadline := time.Now().Add(11 * time.Second)
+	found := false
+	for !found && time.Now().Before(deadline) {
+		node, qm, err := nomad.Nodes().Info(node.ID, qo)
+		must.NoError(t, err)
+		qo.WaitIndex = qm.LastIndex
+		t.Logf("checking node at index %d", qm.LastIndex)
+
+		// This check races with the job scheduling, so only check keyFoo as the
+		// other 2 keys are manipulated by the job itself!
+		if node.Meta[keyFoo] == "bar" {
+			found = true
+		}
+	}
+	must.True(t, found, must.Sprintf("node %q did not update by deadline", node.ID))
+
+	// Wait for the job to complete
+	t.Logf("waiting for job %s to complete", jobID)
+	var alloc *api.AllocationListStub
+	deadline = time.Now().Add(1 * time.Minute)
+	found = false
+	for !found && time.Now().Before(deadline) {
+		allocs, _, err := nomad.Jobs().Allocations(jobID, true, nil)
+		must.NoError(t, err)
+		if len(allocs) > 0 {
+			for _, alloc = range allocs {
+				if alloc.ClientStatus == "complete" {
+					found = true
+					break
+				}
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	must.True(t, found, must.Sprintf("did not find completed alloc"))
```

Comment:
Not for this PR, but there's a `e2eutil.WaitForAllocStatusExpected` which doesn't _quite_ have the API we want I think, because there's a bunch of these little loops all over the e2e tests. Something we should regularize at some point.

Reply:
Agreed, but I also spend a lot of time sprinkling logs and other checks in these loops when iterating on tests.

I think spending some time stuffing those shared funcs full of The Usual Debugging Stuff (eg if an alloc fails, dump Task States and maybe even the last log lines; if a placement fails, dump evals). You may even notice some loops I use WaitIndex while this one I didn't... because it could fail in a way that all of the returned indexes were 0 and handling that was harder than just polling. :grimacing: I think getting some helpers in our *api* package to make some of these common patterns easier to access would payoff for users and tests alike.

I have some more tests I'd like to write post-beta, so we'll see if I make any progress... anything in the *api* package is likely too big of an effort to sneak in though as we really have to make sure to craft the right behavior+api.